### ISSUE TITLE
zernike.arbitrary_basis sometimes clips apertures

### ISSUE THREAD

**Issue by mperrin** [spacetelescope/poppy#221]
<a href="https://github.com/kvangorkom"><img src="https://avatars1.githubusercontent.com/u/9747236?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [kvangorkom](https://github.com/kvangorkom)**
_Tuesday May 16, 2017 at 20:07 GMT_
_Originally opened as https://github.com/mperrin/poppy/issues/221_

----

If passing in an aperture that extends beyond a circle inscribed in the aperture array, `zernike.arbitrary_basis` will clip that aperture. For example, this code...
```
square_aperture = np.zeros((500,500)).astype(bool)
square_aperture[50:-50,50:-50] = 1
square_basis = poppy.zernike.arbitrary_basis(square_aperture)

fig, axes = plt.subplots(1,2,figsize=(10,5))
axes[0].imshow(square_aperture)
axes[0].set_title('Aperture')
axes[1].imshow(square_basis[1])
axes[1].set_title('X-Tilt Term of Arbitrary Basis')
```
...yields this:
![image](https://cloud.githubusercontent.com/assets/9747236/26125475/4362768e-3a50-11e7-8355-a9efc04504db.png)

My current work-around is to compute the smallest circle that will fully enclose the requested aperture and then pad the aperture array so that `zernike.arbitrary_basis` inscribes this circle in the array:

```
# get max extent of aperture
yind, xind = np.where(square_aperture > 0)
distance = np.sqrt( (yind-shape[0]/2.)**2 + (xind-shape[1]/2.)**2 )
max_extent = distance.max()

# pad until circle inscribed in array fully circumscribes aperture
ceil = lambda x: np.ceil(x) if x > 0 else 0
padding = ( int(ceil((max_extent - shape[0] / 2.))),
            int(ceil((max_extent - shape[1] / 2.))) )
padded_aperture = np.pad(square_aperture,padding,mode='constant',constant_values=0)

# create the basis for the padded aperture
padded_basis = poppy.zernike.arbitrary_basis(padded_aperture)

# remove padding
square_basis = [b[padding[0]:-padding[0],padding[1]:-padding[1]] for b in padded_basis]
```
which yields this:
![image](https://cloud.githubusercontent.com/assets/9747236/26125738/357b64b2-3a51-11e7-9688-98187133d706.png)

I can open a PR that adds this to `zernike.arbitrary_basis`, but I figured I'd throw this out there for comment first.

**Comment by mperrin** [comment]
<a href="https://github.com/mperrin"><img src="https://avatars0.githubusercontent.com/u/1151745?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mperrin](https://github.com/mperrin)**
_Wednesday May 17, 2017 at 13:25 GMT_

----

Sure! This all seems reasonable to me. You're absolutely right the Zernike basis needs to start out circumscribing the whole requested aperture, and it makes sense for the function to do that automatically.  I'd say full speed ahead with the PR. 

(I just never worried about this before since the use case which originally motivated this code is the OSIM aperture with all the holes in it, which is neatly circumscribed by a JWST-sized circle...)

**Comment by mperrin** [comment]
<a href="https://github.com/kvangorkom"><img src="https://avatars1.githubusercontent.com/u/9747236?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kvangorkom](https://github.com/kvangorkom)**
_Thursday May 25, 2017 at 15:47 GMT_

----

Woops, forgot about this. Closed by #222!
